### DISCUSSION TITLE
SRI Template Provider as a bitcoin p2p proxy

### DISCUSSION THREAD

**Discussion by Sjors** [stratum-mining/stratum#1036] [Ideas]
I suggested an alternative to the alternative in #1032... https://github.com/bitcoin/bitcoin/pull/29432#issuecomment-2210448369

In this alternative Bitcoin Core would have all the Template Provider functionality, but they would be implemented as p2p messages. The SRI TP would run on the same machine and make a p2p connection with a special `mining` permission so it can access these new message types. It then translates these p2p messages to native Stratum v2.

Just like #1032 this avoids the need to add Noise encryption and a new networking stack to Bitcoin Core. But the SRI side doesn't have to be very smart, it's more like a proxy.

Some downsides are listed in the comment. The main issue is that the Bitcoin p2p protocol is more difficult to use than RPC.

**Comment by GitGab19** [comment]
What would be the real advantage of this solution (against the alternative [here](https://github.com/stratum-mining/stratum/issues/1032)) in your opinion? Could a TP be used by multiple clients in this scenario?

**Reply by Sjors** [reply]
Thanks for moving to discussions.

The advantage would be that the companion (SRI) software can be relatively dumb, just translating Stratum message into bitcoin p2p messages.

But I'm not hugely excited about either approach.

**Comment by Sjors** [comment]
At this point I think the most likely alternative approach to https://github.com/bitcoin/bitcoin/pull/29432 will use IPC, not p2p. So closing this for now.
